### ISSUE TITLE
Implement FLOP counting in GridObjectModel

### ISSUE THREAD

**Issue by hlee9212** [thousandbrainsproject/tbp.floppy#3]
> WOW, what a heroic effort, awesome work! 
> Overall really nice and clean, well documented and love all the tests. Just added a few clarifying comments/suggestions. The main high-level points to highlight are:
> 
> 1) Just a heads up before looking at the comments - my personal preference in PR reviews is if you don't resolve comments, but leave it to me/the reviewer, unless its something trivial like a typo. It helps me keep track of whether something has actually been resolved.
> 2) Re. the learning/training FLOPs, I realized a few things which will impact the results. These differences are:
> - We need to include `self._location_tree = KDTree`, i.e the tree construction, during learning in `GridObjectModel`.
> - We could set `k_n=None` when measuring FLOPs, since we don't actually need the adjacency graph in Monty's inference; this will reduce the number of FLOPs needed for leanring.
> - There is a `rotated_locs = object_rotation.inv().apply(zero_centered_locs)` call during learning - I know we discussed there isn't explicit cover for `apply`, so this would need to be handled manually.
> - I think we might want to increase the K-D Tree construction FLOPs by a factor of 2x (see separate comments). Note this will only impact FLOPs estimated for learning and for inference with voting, but we don't measure the latter for the paper, so only the former matters for the current results.  
> 
> Thanks again for all your hard work!

_Originally posted by @nielsleadholm in https://github.com/thousandbrainsproject/tbp.floppy/pull/2#pullrequestreview-2706756187_
            
Issue to track implementation of `FlopCountingGridObjectModel` and `rotated_locs = object_rotation.inv().apply(zero_centered_locs)`.

**Comment by nielsleadholm** [comment]
Note to self: triple check that our learning configs in DMC actually use `GridObjectModel`, otherwise we can ignore it.

**Comment by nielsleadholm** [comment]
I was a bit confused about where and when K-D tree construction is actually happening in Monty, so I reminded myself of both how the K-D tree is used and constructed for standard (non-voting) evidence updates. Below is a summary as well as some recommendations for the update to Floppy.

## Where is the K-D tree invoked during inference in Evidence-based LMs?

Within `_calculate_evidence_for_new_locations`:
- Initial call to `find_nearest_neighbors`
- This in turn calls `self._location_tree.query`

## Where is this K-D tree (`_location_tree`) constructed during learning?

The answer is not as straightforward as you think, so the following will break this down.

Within `GridObjectModel`, there are the functions `_build_graph_from_grids` and `update_model`, which both have a call:

```python
self._location_tree = KDTree(
    new_graph.pos,
    leafsize=40,
)
```

where the `KDTree` object is imported from `scipy.spatial`

Within `GridObjectModel`, `set_graph` can also instantiate a `KDTree`.

## Why are there three instances of K-D Tree construction in `GridObjectModel`?

If you were training a Monty model with `GridObjectModel` from scratch, i.e. training an `EvidenceGraphLM` (more on this below), then the KDTree would be created for the first time with `_build_graph_from_grids`. After the graph is updated (e.g. object learned at a different rotation), the KDTree would be reconstructed with `update_model`.

**What about `set_graph`?**

`set_graph` is used by `_initialize_model_with_graph` to convert a pre-existing, standard `GraphObjectModel` to the new `GridObjectModel`, i.e. it is invoked by the below snippet within `class EvidenceGraphMemory`:

```python
def _add_graph_to_memory(self, model, graph_id):
        """Add pretrained graph to memory.

        Initializes GridObjectModel and calls set_graph.
        """
    ...
    if isinstance(channel_model, GraphObjectModel):
        # When loading a model trained with a different LM, need to convert
        # it to the GridObjectModel (with use_original_graph == True)
        loaded_graph = channel_model._graph
        channel_model = self._initialize_model_with_graph(
            graph_id, loaded_graph
        )
```

## So which KDTree's are made during learning? The one from scratch, or the conversion one?

`GridObjectModel`s are the main ones we use at inference, since we perform all our evaluations with `EvidenceGraphLM`. In particular, this initializes its graph memory as `self.graph_memory = EvidenceGraphMemory`, which has the following doc string:

```python
class EvidenceGraphMemory(GraphMemory):
    """Custom GraphMemory that stores GridObjectModel instead of GraphObjectModel."""
```

The current setup generally assumes that we do training with a different LM (Displacement), and then evaluate with the `EvidenceGraphMemory`. If we call inference with an `EvidenceGraphLM`, but we performed learning with an `DisplacementGraphLM`, it will invoke the `if isinstance(channel_model, GraphObjectModel):` clause and convert the model to a `GridObjectModel`.

This means that in practice, the conversion process captured in the function `_add_graph_to_memory` is what actually constructs our KD-Tree. As such, the FLOPS associated with K-D tree construction is actually taking place at the start of inference (although at the moment, this is not counted at all). 

For example, our `get_pretrain_lm_config` in DMC defines the LM for learning as follows:

```python
learning_module_class=DisplacementGraphLM
```

This has been the standard practice at the TBP / with Monty because we wanted to store the edges between nodes that `DisplacementGraphLM` creates, in case this would be useful in the future. However, edge-creation is unnecessary FLOPs overhead, and as noted, this approach moves the FLOPs associated with K-D construction (which *is* necessary) into the start of inference, rather than the actual learning phase. 

## Re. voting and K-D Tree:

As an aside, the only other instance of a `KDTree` being created is the `vote_location_tree`, but we are not concerned with voting FLOPs in this repository, so we don't need to make any changes here.

## For FLOPs counting, is there anything other than K-D Tree construction that we might be missing in `GridObjectModel`?
- Their are a few calls to Pytorch in `GridObjectModel`, such as `new_tensor = old_tensor + new_sparse_tensor` and `topk` --> From my understanding @hlee9212 , we can't just plug Monty into `calflops`, and add any FLOPs it counts for Pytorch, to the count we have for Numpy, `KDTree`, right? I believe it expects a Pytorch/`transformers` model?

## Recommendations:

1. For measuring training FLOPs, we should actually perform learning with the `EvidenceGraphLM`. We should probably use the AWS method, given that accommodating all the uses of Pytorch in Floppy would likely take a prohibitive amount of time, unless I am mistaken about the above `calflops` comment. This would mean no KDTree *construction* FLOPs will actually be measured by Floppy, and if taking this approach, we would want to update the main Readme appropriately.
2. For measuring inference FLOPs (which we were going to re-run anyways), we should load the above trained `EvidenceGraphLM` models. The inference accuracy with such a model should be the same as our previous results, since the edges created by `DisplacementGraphLM` are never actually used, but this will enable us to correctly report FLOPs associated with learning vs. inference.
3. Add a note to the Readme to clarify that we don't fully support voting in this version of Floppy.

**Comment by nielsleadholm** [comment]
Per our discussion just now, let's see if it's possible to identify all the relevant Pytorch functions in an hour or two, in which case we can mirror them with Numpy functions per your suggestion. Otherwise we can revert to using the AWS method.

**Comment by nielsleadholm** [comment]
So I spoke to Viviane and it's clearer now that actually the use of the `GridObjectModel` is almost entirely bypassed when we perform inference, because of the argument `use_original_graph=True`. This is very confusing from a code point of view, but basically means that in practice, Evidence LMs are actually using `GraphObjectModel`. The only KD-tree construction calculation takes place when they are initialized at the start of inference.

This means that our approach can be simpler, and basically just requires:

1. Add the KD-tree FLOPs calculation to `set_graph` in `GridObjectModel`, i.e. for the below lines

```python
        if self.use_original_graph:
            # Just use pretrained graph. Do not use grids to constrain nodes.
            self._graph = graph
            self._location_tree = KDTree(
                graph.pos,
                leafsize=40,
            )
```

We don't need to make any changes to `GridObjectModel` other than the above lines, and so we don't need to worry about counting `torch` FLOPs etc in that class. Also note that PPF (point-pair features) are not used in our experiments, so we also don't need to worry about handling their torch-related FLOPs in `GraphObjectModel`.

2. Run learning with a DisplacementLM, i.e. we do *not* need to change the training configs in any fundamental way. The only change we should do is to specify `k_n=None`, so that we don't count FLOPs associated with building the unnecessary edges of the graph. With this change, run experiments to re-measure FLOPs during learning.

3. Re-run FLOPs measurement during inference. The KDTree construction FLOPs will actually be measured during inference, but this shouldn't be an issue, it's basically just a pre-inference step in Monty.


Other notes
- It would be great if, during the above, you can use some logging to determine exactly how much the KD-construction FLOPs actually is, so we know how significant this is compared to other sources. E.g. how often does the construction take place during inference, and how many FLOPs are associated with it.
